basesandbox/src/execution/executor.rs

```diff
@@ -0,0 +1,137 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::ipc::*;
+use std::collections::HashMap;
+use std::process::Command;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+pub trait Executor {
+    fn new(path: &str, args: &[&str]) -> Self;
+    fn join(&mut self);
+}
+
+pub struct Executable {
+    child: std::process::Child,
+}
+
+impl Executor for Executable {
+    fn new(path: &str, args: &[&str]) -> Self {
+        let mut command = Command::new(path);
+        command.args(args);
+        Executable {
+            child: command.spawn().unwrap(),
+        }
+    }
+
+    fn join(&mut self) {
+        // This is synchronized with excutee's signal (#TERMINATE),
+        // which is supposed to be sent in termination step.
+        // Thus, in normal case, it won't take much time to be in a waitable status.
+        // However a malicous excutee might send a signal arbitrarily before the termination.
+        // For that, we have a timeout for the wait.
+        for _ in 0..10 {
+            std::thread::sleep(std::time::Duration::from_millis(50));
+            if self.child.try_wait().unwrap().is_some() {
+                return // successful
+            }
+        }
+        panic!("Module hasn't terminated itself. Protocol Error")
+    }
+}
+
+/// This is for the SameProcess tasks. You can register a runnable function,
+/// and it will be executed later as an instance of 'process' (which is actually not).
+pub type ThreadAsProcesss = Arc<dyn Fn(Vec<String>) -> () + Send + Sync>;
+lazy_static! {
+    static ref POOL: Mutex<HashMap<String, ThreadAsProcesss>> = { Mutex::new(HashMap::new()) };
+}
+
+pub fn add_plain_thread_pool(key: String, f: ThreadAsProcesss) {
+    POOL.lock().unwrap().insert(key, f);
+}
+
+pub struct PlainThread {
+    handle: Option<std::thread::JoinHandle<()>>,
+}
+
+impl Executor for PlainThread {
+    fn new(path: &str, args: &[&str]) -> Self {
+        let path = path.to_owned();
+        let mut args: Vec<String> = args.iter().map(|&x| x.to_string()).collect();
+        args.insert(0, "Thread".to_owned()); // corresponding to program path
+        let handle = std::thread::spawn(move || {
+            let f = POOL.lock().unwrap().get(&path).unwrap().clone();
+            f(args)
+        });
+
+        PlainThread {
+            handle: Some(handle),
+        }
+    }
+
+    fn join(&mut self) {
+        // PlainThread Executor is for test, so no worry for malicous unresponsiveness
+        self.handle.take().unwrap().join().unwrap();
+    }
+}
+
+/// Rust doesn't allow Drop for trait, so we need this. See E0120
+struct ExecutorWrapper<T: Executor> {
```

Comment:
Why do we need a wrapper here?
```rust
trait Executor {
    fn join(&self) {}
}


struct A;

impl Drop for A {
    fn drop(&mut self) {
        self.join();
    }
}

impl Executor for A {
    fn join(&self) {}
}
```

???

Reply:
@dynaxis I have to implement `Drop` for N time, for `struct A`, `struct B`, `struct C`, ...